### PR TITLE
Adding whitelistHandler for supporting yaml version of whitelist file

### PR DESCRIPTION
**Pull request by shigupt202**
Adding support for yaml version of whitelist file with following format:
image

We can change the label names based on PM input later.

### CODE REVIEW

File: action.yaml [hunk 1]
```diff
@@ -4,21 +4,15 @@ inputs:
...
     description: 'severities of vulnerabilities to be displayed (UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL)'
-    reuired: false
+    required: false
   username:
     description: 'Username to authenticate to the Docker registry'
     required: false
   password:
     description: 'Password to authenticate to the Docker registry'
     required: false
-  github-token:
```
Comment (ajinkya599) [inline action.yaml:19]: Revert this. We need this token for creating check-run.

File: package.json [hunk 2]
```diff
@@ -16,7 +16,9 @@
   "dependencies": {
     "@actions/core": "^1.2.3",
     "@actions/tool-cache": "^1.3.1",
+    "@types/js-yaml": "^3.12.2",
```
Comment (ajinkya599) [inline package.json]: This should not be required?

File: src/inputHelper.ts [hunk 3]
```diff
@@ -1,9 +1,7 @@
 import * as core from '@actions/core';
 
 export const imageName = core.getInput("image-name");
-export const githubToken = core.getInput("github-token");
```
Comment (ajinkya599) [inline src/inputHelper.ts:4]: Revert this.

File: src/main.ts [hunk 4]
```diff
@@ -1,31 +1,18 @@
...
-}
+import * as whitelistHandler from './whitelistHandler';
 
 async function getTrivyEnvVariables(): Promise<{ [key: string]: string }> {
     let trivyEnv: { [key: string]: string } = {};
     for (let key in process.env) {
         trivyEnv[key] = process.env[key] || "";
     }
 
-    const githubToken = inputHelper.githubToken;
```
Comment (ajinkya599) [inline src/main.ts:26]: Revert this as well, since we are taking the github-token input for check run creation.

File: src/whitelistHandler.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    trivyWhitelistPath = trivyWhitelistDir + "/whitelist";
+}
+
+function initializeDockleWhitelistPath() {
+    //Dockle expects .dockleignore file at the root of the repo 
+    dockleWhitelistPath = `${process.env['GITHUB_WORKSPACE']}/.dockleignore`;
+}
+
+export function init() {
+    const whitelistFilePath = `${process.env['GITHUB_WORKSPACE']}/.github/containerscan/whitelist`;
```
Comment (ajinkya599) [inline src/whitelistHandler.ts]: Add extension `.yaml` to the file?

File: src/whitelistHandler.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+function initializeTrivyWhitelistPath() {
+    let curDate = Date.now();
+    let trivyWhitelistDir = `${process.env['GITHUB_WORKSPACE']}/containerscan_${curDate}/trivy`;
+    fs.mkdirSync(trivyWhitelistDir, { recursive: true });
+    trivyWhitelistPath = trivyWhitelistDir + "/whitelist";
+}
+
+function initializeDockleWhitelistPath() {
+    //Dockle expects .dockleignore file at the root of the repo 
+    dockleWhitelistPath = `${process.env['GITHUB_WORKSPACE']}/.dockleignore`;
```
Comment (ajinkya599) [inline src/whitelistHandler.ts:24]: Can't we keep it `${process.env['GITHUB_WORKSPACE']}/containerscan_${curDate}/.dockleignore` and pass `cwd` in `options` while instantiating `ToolRunner`?

File: src/whitelistHandler.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+export function getDockleWhitelist(): string {
+    if (dockleWhitelistExists)
+        return dockleWhitelistPath;
+    else
+        throw new Error("Could not find whitelist file for best practice vulnerabilities");
+}
+
+function initializeTrivyWhitelistPath() {
+    let curDate = Date.now();
+    let trivyWhitelistDir = `${process.env['GITHUB_WORKSPACE']}/containerscan_${curDate}/trivy`;
```
Comment (ajinkya599) [inline src/whitelistHandler.ts:17]: Name it `.trivyignore` ?

File: src/whitelistHandler.ts (hunk 6 above)
Comment (ajinkya599) [reply src/whitelistHandler.ts:24]: We should use a single date stamp for all the resources associated with this run.
Refer to `fileHelper` in this [PR](https://github.com/shigupt202/local-container-scan/pull/6/files). You can use `getContainerScanDirectory()` once it's checked-in.

File: src/whitelistHandler.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+export function init() {
+    const whitelistFilePath = `${process.env['GITHUB_WORKSPACE']}/.github/containerscan/whitelist`;
+    if (!fs.existsSync(whitelistFilePath)) {
+        console.log("Could not find whitelist file.");
+        return;
+    }
+    initializeTrivyWhitelistPath();
+    initializeDockleWhitelistPath();
```
Comment (ajinkya599) [inline src/whitelistHandler.ts:35]: knit: keep some space above and below.

File: src/whitelistHandler.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+export let dockleWhitelistExists = false;
+
+export function getTrivyWhitelist(): string {
+    if (trivyWhitelistExists)
+        return trivyWhitelistPath;
+    else
+        throw new Error("Could not find whitelist file for common vulnerabilities");
+}
+
+export function getDockleWhitelist(): string {
```
Comment (ajinkya599) [inline src/whitelistHandler.ts]: Is this used anywhere?

File: src/whitelistHandler.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                let dockleWhitelistContent = "";
+                vulnArray.forEach(vuln => {
+                    dockleWhitelistContent += vuln;
+                    dockleWhitelistContent += "\n";
+                });
+                fs.writeFileSync(dockleWhitelistPath, dockleWhitelistContent);
+            }
+        }
+    } catch (error) {
+        throw new Error("Error while loading whitelist file")
```
Comment (ajinkya599) [inline src/whitelistHandler.ts]: Change to "parsing" and log the actual error -
Error while **parsing** the whitelist file. Error: error

File: src/whitelistHandler.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+export function init() {
+    const whitelistFilePath = `${process.env['GITHUB_WORKSPACE']}/.github/containerscan/whitelist`;
+    if (!fs.existsSync(whitelistFilePath)) {
+        console.log("Could not find whitelist file.");
+        return;
+    }
+    initializeTrivyWhitelistPath();
+    initializeDockleWhitelistPath();
+    try {
+        var whitelist_yaml = jsyaml.safeLoad(fs.readFileSync(whitelistFilePath, 'utf8'));
```
Comment (ajinkya599) [inline src/whitelistHandler.ts]: `const`

File: src/whitelistHandler.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        if (whitelist_yaml.general) {
+            if (whitelist_yaml.general.common_vulnerabilities) {
+                trivyWhitelistExists = true;
+                let vulnArray: string[] = whitelist_yaml.general.common_vulnerabilities;
```
Comment (ajinkya599) [inline src/whitelistHandler.ts]: `const`

File: src/whitelistHandler.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                let trivyWhitelistContent = "";
+                vulnArray.forEach(vuln => {
+                    trivyWhitelistContent += vuln;
+                    trivyWhitelistContent += "\n";
+                });
+                fs.writeFileSync(trivyWhitelistPath, trivyWhitelistContent);
+            }
+            if (whitelist_yaml.general.best_practice_vulnerabilities) {
+                dockleWhitelistExists = true;
+                let vulnArray: string[] = whitelist_yaml.general.best_practice_vulnerabilities;
```
Comment (ajinkya599) [inline src/whitelistHandler.ts]: `const`

File: src/main.ts [hunk 14]
```diff
@@ -85,12 +66,11 @@ async function setDockleEnvVariables(): Promise<{ [key: string]: string }> {
         dockleEnv["DOCKLE_PASSWORD"] = password;
     }
 
-    dockleEnv["DOCKLE_EXIT_CODE"] = "5";
-
     return dockleEnv;
 }
 
 async function runTrivy(): Promise<number> {
+    whitelistHandler.init();
```
Comment (ajinkya599) [inline src/main.ts]: This initialises both Trivy and Dockle whitelists. Move this call to `run`.

File: src/whitelistHandler.ts (hunk 5 above)
Comment (ajinkya599) [reply src/whitelistHandler.ts]: Shouldn't it be `${process.env['GITHUB_WORKSPACE']}/${repoName}/.github/containerscan/whitelist`?
From the [doc](https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables) it seems to be the case.

File: src/whitelistHandler.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    if (!fs.existsSync(whitelistFilePath)) {
+        console.log("Could not find whitelist file.");
+        return;
+    }
+    initializeTrivyWhitelistPath();
+    initializeDockleWhitelistPath();
+    try {
+        var whitelist_yaml = jsyaml.safeLoad(fs.readFileSync(whitelistFilePath, 'utf8'));
+        if (whitelist_yaml.general) {
+            if (whitelist_yaml.general.common_vulnerabilities) {
```
Comment (ajinkya599) [inline src/whitelistHandler.ts]: Should we use camelCase - `commonVulnerabilities` ? (Kubernetes uses camelCase. eg. `apiVersion`, `restartPolicy` etc. I am thinking of uniformity with the eco-system)

**Review by ajinkya599 — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (shigupt202) [reply package.json]: It got added automatically when I installed the module. It's needed for parsing the whitelist file. Should I remove it?

File: action.yaml (hunk 1 above)
Comment (shigupt202) [reply action.yaml:19]: Okay

File: src/whitelistHandler.ts (hunk 5 above)
Comment (shigupt202) [reply src/whitelistHandler.ts]: I'll add the .yaml extension. Regarding the file path, repo name is not required. GITHUB_WORKSPACE contains it already. I've tested with this path, it works.

File: src/whitelistHandler.ts (hunk 6 above)
Comment (shigupt202) [reply src/whitelistHandler.ts:24]: Okay

File: src/whitelistHandler.ts (hunk 7 above)
Comment (shigupt202) [reply src/whitelistHandler.ts:17]: That's the folder name. I'll change the file name to .trivyignore

File: src/whitelistHandler.ts (hunk 9 above)
Comment (shigupt202) [reply src/whitelistHandler.ts]: No. I'll remove it.

File: src/whitelistHandler.ts (hunk 15 above)
Comment (shigupt202) [reply src/whitelistHandler.ts]: Okay. I'll change that

File: src/main.ts (hunk 4 above)
Comment (shigupt202) [reply src/main.ts:26]: Do we need github token in this function for check run creation?

File: src/whitelistHandler.ts (hunk 10 above)
Comment (shigupt202) [reply src/whitelistHandler.ts]: throw new Error("Error while parsing whitelist file. Error: " + error); Something like this?

File: src/main.ts (hunk 14 above)
Comment (shigupt202) [reply src/main.ts]: OK

File: src/main.ts (hunk 4 above)
Comment (ajinkya599) [reply src/main.ts:26]: This is to increase throttling limits as you explained earlier.

File: action.yaml (hunk 1 above)
Comment (thesattiraju) [reply action.yaml:19]: Looks like you just moved it to a different line

File: package.json (hunk 2 above)
Comment (thesattiraju) [reply package.json]: `@types` can be a part of devDependencies

File: src/whitelistHandler.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        console.log("Could not find whitelist file.");
+        return;
+    }
+
+    initializeTrivyWhitelistPath();
+    initializeDockleWhitelistPath();
+
+    try {
+        const whitelistYaml = jsyaml.safeLoad(fs.readFileSync(whitelistFilePath, 'utf8'));
+        if (whitelistYaml.general) {
```
Comment (thesattiraju) [inline src/whitelistHandler.ts:39]: What's `.general` for? Why are we forking it like this?

File: src/whitelistHandler.ts [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+                });
+                fs.writeFileSync(trivyWhitelistPath, trivyWhitelistContent);
+            }
+            if (whitelistYaml.general.bestPracticeVulnerabilities) {
+                const vulnArray: string[] = whitelistYaml.general.bestPracticeVulnerabilities;
+                let dockleWhitelistContent = "";
+                vulnArray.forEach(vuln => {
+                    dockleWhitelistContent += vuln;
+                    dockleWhitelistContent += "\n";
+                });
```
Comment (thesattiraju) [inline src/whitelistHandler.ts:56]: Use `.join` function instead

File: src/whitelistHandler.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        const whitelistYaml = jsyaml.safeLoad(fs.readFileSync(whitelistFilePath, 'utf8'));
+        if (whitelistYaml.general) {
+            if (whitelistYaml.general.commonVulnerabilities) {
+                trivyWhitelistExists = true;
+                const vulnArray: string[] = whitelistYaml.general.commonVulnerabilities;
+                let trivyWhitelistContent = "";
+                vulnArray.forEach(vuln => {
+                    trivyWhitelistContent += vuln;
+                    trivyWhitelistContent += "\n";
+                });
```
Comment (thesattiraju) [inline src/whitelistHandler.ts:47]: Use `.join` function instead

File: src/whitelistHandler.ts (hunk 16 above)
Comment (thesattiraju) [reply src/whitelistHandler.ts:39]: You don't seem to be handling it for other images. Is this expected? Based on the image you've pasted in description

**Review by ajinkya599 — APPROVED** [review]

File: src/whitelistHandler.ts (hunk 16 above)
Comment (shigupt202) [reply src/whitelistHandler.ts:39]: For the first version of the action, we are supporting only a single list of vulnerabilities for CVEs and CISs. In later versions, we are planning to support image specific whitelisting. Keeping in mind back-compatibility and extensibility, I've added the general label where the customer can put all the vulnerabilities irrespective of the image.

File: src/whitelistHandler.ts (hunk 17 above)
Comment (shigupt202) [reply src/whitelistHandler.ts:56]: strings do not seem to support join function. It's probably used with arrays.

File: src/whitelistHandler.ts (hunk 8 above)
Comment (shigupt202) [reply src/whitelistHandler.ts:35]: Done

File: action.yaml (hunk 1 above)
Comment (shigupt202) [reply action.yaml:19]: Updated

File: src/whitelistHandler.ts (hunk 5 above)
Comment (shigupt202) [reply src/whitelistHandler.ts]: Tracking the issue for .yml extension [here](https://github.com/shigupt202/local-container-scan/issues/8)

File: src/whitelistHandler.ts (hunk 6 above)
Comment (shigupt202) [reply src/whitelistHandler.ts:24]: Tracking the issue for a specific path for .dockleignore [here](https://github.com/shigupt202/local-container-scan/issues/8)

File: src/whitelistHandler.ts (hunk 17 above)
Comment (thesattiraju) [reply src/whitelistHandler.ts:56]: You could have used it like `vulnArray.join("\n")`

Comment (shigupt202) [reply src/whitelistHandler.ts:56]: Will do that in next PR for output formatting.

File: package.json (hunk 2 above)
Comment (shigupt202) [reply package.json]: Moved

**Pull request merged**